File: LeetCode/Dropbox/200_NumIslands.py
import collections

class Solution:
    def queue(self, grid, i, j):
        queue = collections.deque([])
        queue.append((i,j))
        while queue:
            ii, jj = queue.pop()
            if 0 <= ii < len(grid) and 0 <= jj < len(grid[0]) and grid[ii][jj] == '1':
                grid[ii][jj] = '0'
                queue.append((ii-1, jj))
                queue.append((ii+1, jj))
                queue.append((ii, jj-1))
                queue.append((ii, jj+1))


    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0

        n = len(grid)
        m = len(grid[0])

        islands = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == '1':
                    islands += 1
                    self.queue(grid, i, j)

        return islands

File: LeetCode/Dropbox/200_NumIslands.cs
namespace LeetCode.Dropbox
{
    public class NumIslands
    {
        public int NumIslands1(char[][] grid)
        {
            if (grid == null || grid.Length == 0 || grid[0].Length == 0)
            {
                return 0;
            }

            int n = grid.Length;
            int m = grid[0].Length;

            int islands = 0;
            for (int i = 0; i < n; i++)
            {
                for (int j = 0; j < m; j++)
                {
                    if (grid[i][j] == '1')
                    {
                        islands++;
                        check(grid, i, j);
                    }
                }
            }

            return islands;
        }

        private void check(char[][] grid, int n, int m)
        {
            if (n < 0 || m < 0 || n >= grid.Length || m >= grid[0].Length || grid[n][m] == '0') return;

            grid[n][m] = '0';

            check(grid, n - 1, m);
            check(grid, n + 1, m);
            check(grid, n, m - 1);
            check(grid, n, m + 1);
        }
    }
}
